libs/x-content/src/main/java/org/elasticsearch/common/xcontent/MediaTypeParser.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.xcontent;
+
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+
+public class MediaTypeParser<T extends MediaType> {
+    private Map<String, T> formatToMediaType = new HashMap<>();
+    private Map<String, T> typeSubtypeToMediaType = new HashMap<>();
+
+    public MediaTypeParser(T[] acceptedMediaTypes) {
+        for (T mediaType : acceptedMediaTypes) {
+            typeSubtypeToMediaType.put(mediaType.typeSubtype(), mediaType);
+            formatToMediaType.put(mediaType.format(), mediaType);
+        }
+    }
+
+    public T fromMediaType(String mediaType) {
+        ParsedMediaType parsedMediaType = parseMediaType(mediaType);
+        return parsedMediaType != null ? parsedMediaType.getMediaType() : null;
+    }
+
+    public T fromFormat(String format) {
+        if(format == null) {
+            return null;
+        }
+        return formatToMediaType.get(format.toLowerCase(Locale.ROOT));
+    }
+
+    public MediaTypeParser<T> withAdditionalMediaType(String typeSubtype, T xContentType) {
+        typeSubtypeToMediaType.put(typeSubtype.toLowerCase(Locale.ROOT), xContentType);
+        formatToMediaType.put(xContentType.format(), xContentType);
+        return this;
+    }
+
+    public ParsedMediaType parseMediaType(String mediaType) {
+        if (mediaType != null) {
+            String headerValue = mediaType.toLowerCase(Locale.ROOT);
+            // split string on semicolon
+            // validate media type is accepted (IIRC whitespace is ok so trim it) //TODO PG whitespace only ok in params
+            // rest of strings are params. validate per RFC 7230 and use ones that we care about
+            // or use a regex and we can change if necessary
+            String[] split = headerValue.split(";");
+
+            String[] typeSubtype =  split[0].toLowerCase(Locale.ROOT)
+                                .split("/");
+            if (typeSubtype.length == 2) {
```

Comment:
4x yes.
so this is exactly the intention here. 
re1 The parseMediaType is not supporting the format parsing.
re2 The caller has to know which one to call.
re3 There should not be a place were we parse a format but a value is taken from Accept or Content-Type headers. Values from headers should follow media-type format (type/subtype;parameters)
re4 I hope after this PR this will be clear that mime types are for headers, format for query parameter

Reply:
thanks.. i double checked `TextTemplateEngine` and while it appears that the former supported either application/json or json,  I think we are fine (as-is here) to only support json for these reasons:   1) only applicable to inline mustache scripts in watcher 2) it is not documented and only lightly tested 3) only found one example of its usage in the wild  (and it used json).  Given the lack of documentation i doubt this is used much at all and if so likely uses the short form `json`.

I have not validated `RestSqlQueryAction` behavior

Reply:
`RestSqlQueryAction` was reviewed by @bpintea